src/nr_openai_observability/bedrock.py

```diff
@@ -0,0 +1,414 @@
+        )
+    except AttributeError as error:
+        logger.debug(f'failed to instrument botocore.client.ClientCreator.create_client: {error}')
+
+
+def patcher_aws_create_api(original_fn, *args, **kwargs):
+    # We are in the AWS API to create a new object, we need to invoke that first!
+    try:
+        from botocore.model import ServiceModel
+        response = original_fn(*args, **kwargs)
+
+        cls = type(response)
+        name = cls.__qualname__
+        if cls.__module__ is not None and cls.__module__ != '__builtin__':
+            name = f'{cls.__module__}.{cls.__qualname__}'
+
+        if name == 'botocore.client.BedrockRuntime':
+            bedrock_method = 'invoke_model'
+            original_invoke_model = getattr(response, bedrock_method)
+
+            if original_invoke_model is not None:
+                setattr(response, bedrock_method, _patched_call(
+                    original_invoke_model, 
+                    patcher_bedrock_create_completion
+                ))
+            else:
+                logger.error(f"failed to find method '{bedrock_method}' on {response}")
+
+        return response
+
+    except BaseException as error:
+        logger.error(f'caught exception trying to invoke original function: {error}')
+        raise error
+
+
+def patcher_bedrock_create_completion(original_fn, *args, **kwargs):
+    timestamp = time.time()
+    result, time_delta = None, None
+
+    try:
+        with newrelic.agent.FunctionTrace(
+            name="AI/Bedrock/Chat/Completions/Create", group="", terminal=True
+        ):
+            monitor.record_library('botocore', 'Bedrock')
+            result = original_fn(*args, **kwargs)
+            time_delta = time.time() - timestamp
+    except Exception as error:
+        build_completion_summary_for_error(error, **kwargs)
+        raise result
+
+    # print the HTTP body
+    from botocore.response import StreamingBody
+    from io import BytesIO
+
+    contents = result['body'].read()
+    bio = BytesIO(contents)
+    result['body'] = StreamingBody(bio, len(contents))
+
+    handle_bedrock_create_completion(result, time_delta, **kwargs)
+
+    # we have to reset the body after we read it. The handle function is going to read the contents,
+    # and we'll have to apply the body again before returning back.
+    bio = BytesIO(contents)
+    result['body'] = StreamingBody(bio, len(contents))
+    return result
+
+
+def build_completion_summary_for_error(error, **kwargs):
+    logger.error(f"error invoking bedrock function: {error}")
+
+    completion = {
+        "id": str(uuid.uuid4()),
+        "vendor": "bedrock",
+        "ingest_source": "PythonSDK",
+        "timestamp": datetime.now(),
+        "error_status": error.http_status,
+        "error_message": error.error.message,
+        "error_type": error.error.type,
+        "error_code": error.error.code,
+        "error_param": error.error.param,
+    }
+
+    monitor.record_event(completion, EventName)
```

Comment:
Should this be the SummaryEvent vs the Completion Event?

Reply:
Yes it should be. Done.